### PR TITLE
Print the Memo in topological order.

### PR DESCRIPTION
**Pull request by albler**
It is easier to navigate the Memo visually if the plans show up as inverted
trees. This change adds (a quick and dirty) topological sort for groups of the
Memo and uses it in that class's Stringer.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -181,10 +182,96 @@ func (m *memo) list(n *node) {
...
+type byFDesc []*dfsInfo
+
+func (a byFDesc) Len() int           { return len(a) }
+func (a byFDesc) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byFDesc) Less(i, j int) bool { return a[i].F >= a[j].F }
+
+func (m *memo) topoSort() []*class {
+
+	dfs := m.DFS()
+	sort.Sort(byFDesc(dfs))
```
Comment (petermattis) [inline main.go]: FYI, go1.8 supports `sort.Slice`:

```
sort.Slice(dfs, func(i, j int) bool {
  return a[i].F >= a[j].F
})
```

Not quite as fast as `sort.Sort`, but speed isn't important here.

File: main.go [hunk 2]
```diff
@@ -181,10 +182,96 @@ func (m *memo) list(n *node) {
...
+		}
+		res = append(res, state[c])
+	}
+	return res
+}
+
+func (m *memo) dfsVisit(c *class, t *int, state map[*class]*dfsInfo) {
+	*t++
+	state[c].D = *t
+	state[c].Color = gray
```
Comment (petermattis) [inline main.go]: I haven't implemented topological sort in a long time, but I seem to recall only having 2 states: visited and not visited. Perhaps I'm misremembering.

Comment (albler) [inline main.go]: There's a feature that I suppressed here in which you classify the edges that you traverse, e.g., back edges. The gray state is used in that classification. I quickly pulled this implementation from an old repo of mine but stripped edges classification. The assumption here is we'll have trees anyway.

File: main.go (hunk 1 above)
Comment (albler) [inline main.go]: Agreed. It cuts a few lines on an area that's not really our focus now. I'll change.

File: main.go (hunk 2 above)
Comment (petermattis) [inline main.go]: Ack.

File: main.go [hunk 3]
```diff
@@ -181,10 +182,90 @@ func (m *memo) list(n *node) {
...
+		}
+	}
+
+	state[c].Color = black
+	*t++
+	state[c].F = *t
+}
+
+func (m *memo) topoSort() []*class {
+
```
Comment (petermattis) [inline main.go]: Spurious blank line which my OCD will likely remove at some point.

**Pull request merged**